### ISSUE TITLE
100x Performance regression with @fastmath and combination of two Units

### ISSUE THREAD

**Issue by zsoerenm** [PainterQubits/Unitful.jl#281]
Here is MWE:
```julia
function accumulate_phase(f0, fs)
    phase_accu = 0.0
    @fastmath for i = 1:4000
        phase_accu += f0 * i / fs
    end
    phase_accu
end
```
```julia
julia> @btime accumulate_phase($(100.23Hz), $(4MHz))
  249.836 μs (0 allocations: 0 bytes)
200.510115

julia> @btime accumulate_phase($(100.23Hz), $(4e6Hz))
  2.792 μs (0 allocations: 0 bytes)
200.51011499999998

julia> @btime accumulate_phase($(100Hz), $(4MHz))
  3.303 μs (0 allocations: 0 bytes)
200.05
```
That is **249.836 μs** against 2.792 μs

Without fastmath:
```julia
function accumulate_phase_wo_fastmath(f0, fs)
    phase_accu = 0.0
    for i = 1:4000
        phase_accu += f0 * i / fs
    end
    phase_accu
end
```
```julia
julia> @btime accumulate_phase_wo_fastmath($(100.23Hz), $(4MHz))
  4.489 μs (0 allocations: 0 bytes)
200.510115

julia> @btime accumulate_phase_wo_fastmath($(100.23Hz), $(4e6Hz))
  4.458 μs (0 allocations: 0 bytes)
200.510115
```

**Comment by ajkeller34** [comment]
This is kind of a fun one. To resolve your immediate issue, I would suggest you just call `float` on your input arguments. `phase_accu` is a floating-point number anyway, so this should work alright for you. 

The reason this happens is because unit conversion for integer and rational `Quantity`s is done exactly. The regression you found happens during promotion: a result `Quantity` type is found, and the unit conversion happens exactly for `4MHz` before the result is made into a floating-point representation. The offending line is [here](https://github.com/PainterQubits/Unitful.jl/blob/7ff5c0d7cbdce4500c6f96d28b7b8eec061380f9/src/conversion.jl#L106). Because the conversion involves rationals, it is very slow, even though the fast math implementation is being used on the resulting floating-point `Quantity`s.

It's been a while since I thought about this, but I think the current implementation is motivated by cases like this:

```
julia> convert(typeof(0.1u"m"), (0.2)u"dm")
0.020000000000000004 m

julia> convert(typeof(0.1u"m"), (2//10)u"dm")
0.02 m
```

In other words, the logic is that if the Quantity you start with is represented exactly using a rational or integer `Quantity`, and the conversion factors can also be expressed exactly, then there should be no error permitted in the conversion, even if converting to a floating-point `Quantity`. I suppose it would be possible to write some logic to override this behavior for fastmath only but it is not so clear that really makes sense—I think the solution is probably just to ensure floats are passed in. I'm not saying that this is an obvious problem though. We could certainly add some documentation to warn users about this.

**Comment by zsoerenm** [comment]
If the offending line is [https://github.com/PainterQubits/Unitful.jl/blob/7ff5c0d7cbdce4500c6f96d28b7b8eec061380f9/src/conversion.jl#L106](here). Why does the performance regression only occure in the case of `@fastmath`?

**Comment by ajkeller34** [comment]
The implementation of `@fastmath` has a fallback that uses promotion (and hence unit conversion) if the numeric types `T` in `Quantity{T}` are not equal. If you don't use `@fastmath`, no unit conversion happens until you need to add the right-hand side to the left, at which point the right-hand side is already a floating-point number and unit conversion is fast. My implementation of `@fastmath` for `Quantity`s basically mirrors what is in Base but it could be that there is a better way to do it here.

**Comment by cmichelenstrofer** [comment]
> We could certainly add some documentation to warn users about this

That would be great! Exact conversion is a nice feature but it would be good to mention that it comes at a performance cost. And mention the "trick" of passing in floats. I don't think this is specific to `fastmath`. We saw significant slowdown when using trig functions with integer vs float: yakir12/UnitfulAngles.jl#23.
